README.md
# cb-tomcat

Configures a Tomcat 7 installation. 

- SCM: https://github.com/frkline/cb-tomcat
- Build: 

Note that Tomcat 7 has been chosen as tools support for Tomcat 8 is currently limited. For instance, the Maven plugin to remote deploy WARs to Tomcat does not currently support Tomcat 8: https://issues.apache.org/jira/browse/MTOMCAT-234.

For more information, visit [cb-tomcat](https://github.com/frkline/cb-tomcat)

## Features

- Installs Tomcat 7
- Configure JMX bind and ephemeral ports, address, and local port enforcement with [JmxRemoteLifecycleListener](http://tomcat.apache.org/tomcat-7.0-doc/config/listeners.html#JMX_Remote_Lifecycle_Listener_-_org.apache.catalina.mbeans.JmxRemoteLifecycleListener)
- Configure [JDWP](http://docs.oracle.com/javase/1.5.0/docs/guide/jpda/jdwp-spec.html) for remote debugging during development
- Configure manager-script role for access to the [Manager Application](http://tomcat.apache.org/tomcat-7.0-doc/manager-howto.html#Configuring_Manager_Application_Access) from the [Maven Tomcat 7 Plugin](http://tomcat.apache.org/maven-plugin-2.0/tomcat7-maven-plugin/)

## Supports

- CentOS 6.5

## Configure

```ruby
# Basic installation
node.default['tomcat']['user'] = 'tomcat'
node.default['tomcat']['group'] = 'tomcat'
node.default['tomcat']['version'] = '7.0.54'
node.default['tomcat']['install_directory'] = '/opt'
node.default['tomcat']['java_opts'] = [
  '-Djava.awt.headless=true',
  '-Dfile.encoding=UTF-8',
  '-server'
]
node.default['tomcat']['catalina_opts'] = [
]

# Manager
node.default['tomcat']['webapps']['manager']['deploy'] = true
node.default['tomcat']['webapps']['manager']['admin-script']['enabled'] = true
node.default['tomcat']['webapps']['manager']['admin-script']['pw'] = 'script'
node.default['tomcat']['webapps']['manager']['admin-gui']['enabled'] = true
node.default['tomcat']['webapps']['manager']['admin-gui']['pw'] = 'gui'

# JMX
default['tomcat']['jmx']['rmi_registry_port'] = '10001'
default['tomcat']['jmx']['rmi_server_port'] = '10002'
default['tomcat']['jmx']['rmi_bind_address'] = 'localhost'
default['tomcat']['jmx']['use_local_ports'] = false
default['tomcat']['jmx']['ssl'] = false
default['tomcat']['jmx']['authenticate'] = true
default['tomcat']['jmx']['password_file'] =
  '$CATALINA_HOME/conf/jmxremote.password'
default['tomcat']['jmx']['access_file'] =
  '$CATALINA_HOME/conf/jmxremote.access'
default['tomcat']['jmx']['username'] = 'admin-jmx'
default['tomcat']['jmx']['password'] = 'jmx'

# JDWP
default['tomcat']['jdwp']['enabled'] = true
default['tomcat']['jdwp']['suspend'] = false
default['tomcat']['jdwp']['port'] = '10003'

# server.xml
default['tomcat']['shutdown_port'] = '8005'
default['tomcat']['shutdown'] = 'shutdown'
default['tomcat']['executor']['max_threads'] = '100'
default['tomcat']['executor']['min_spare_threads'] = '10'
default['tomcat']['connector']['http']['protocol'] =
  'org.apache.coyote.http11.Http11NioProtocol'
default['tomcat']['connector']['http']['port'] = '8080'
default['tomcat']['connector']['http']['redirect_port'] = '8443'
default['tomcat']['connector']['http']['connection_timeout'] = '20000'
default['tomcat']['connector']['http']['uri_encoding'] = 'UTF-8'
```

## Usage

Prerequisite: [Configure the Cookbook](#configure-the-cookbook)

**Start a CentOS 6.5 VirtualBox VM**
```
> cd cookbook
> bundle exec vagrant up
```

**Connect to the VM**
```
> cd cookbook
> bundle exec vagrant ssh
```

## Development

0. [Configure Your Environment](https://github.com/frkline/dev-setup/#configure-your-environment)
1. Clone the Repository  

  ```
  > git clone --recursive git@github.com:frkline/cb-tomcat.git
  ```  
   
2. Initialize the Cookbook's Dependencies  

  ```
  > cd cb-common/cookbook  
  > bundle install
  ```

  Note: Upon update of the Gemfile, update the Bundle:
  ```
  > cd cookbook
  > bundle install
  ```  
  
3. Start Guard  

  Guard will test, analyze, and lint the cookbook as changes are made by monitoring
  your local directory for changes. It will run ChefSpec, Foodcritic, Rubocop, and ServerSpec tests
  as required. For more information, see: https://github.com/test-kitchen/guard-kitchen.
  ```
  > cd cookbook
  > bundle exec guard start
  ```
  
### Development Tools

**Run ChefSpec**  
ChefSpec is a unit testing framework for testing Chef cookbooks. ChefSpec makes it easy to write examples and get fast feedback on cookbook changes without the need for virtual machines or cloud servers. For more information, see: https://github.com/sethvargo/chefspec.
```
> cd cookbook
> bundle exec rspec
```

**Run Foodcritic**  
Foodcritic is a lint tool for Opscode Chef cookbooks. Along with the default checks, we include community-standard checks from [Etsy](https://github.com/etsy/foodcritic-rules) as well as [CustomInk](https://github.com/customink-webops/foodcritic-rules) For more information, see: http://acrmp.github.io/foodcritic/.
```
> cd cookbook
> bundle exec foodcritic -I foodcritic/* .
```

**Run RuboCop**  
RuboCop is a static code analyzer, based upon the standard Ruby community style guide. For more information, see: https://github.com/bbatsov/rubocop.
```
> cd cookbook
> bundle exec rubocop
```

**Run Tests with Kitchen**  
Kitchen is an integration tool for developing and testing infrastructure code and software on isolated target platforms. Used alongside ServerSpec, we can write RSpec tests for checking if servers will be configured properly. For more information, see: https://github.com/test-kitchen/test-kitchen and http://serverspec.org/.
```
> cd cookbook
> bundle exec kitchen verify
```



CHANGELOG.md
# 0.1.0

Initial release of cb-tomcat


cookbook/spec/tomcat_spec.rb
require 'chefspec'
require_relative 'spec_helper'

describe 'cb-tomcat::default' do
  let(:chef_run) { ChefSpec::Runner.new.converge(described_recipe) }

  it 'includes the java::default recipe' do
    expect(chef_run).to include_recipe('java::default')
  end

  it 'includes the cb-tomcat::tomcat recipe' do
    expect(chef_run).to include_recipe('cb-tomcat::tomcat')
  end

end


cookbook/test/integration/default/serverspec/tomcat_spec.rb
require 'spec_helper'

# Java
describe command('java -version') do
  its(:stdout) do
    should match(/#{Regexp.escape("java version \"1.8.0")}/)
  end
  its(:stdout) do
    should match(/#{Regexp.escape('(TM) SE Runtime Environment (build 1.8.0')}/)
  end
  its(:stdout) do
    should match(/#{Regexp.escape('Java HotSpot(TM) 64-Bit Server VM')}/)
  end
end

# Tomcat
describe service('tomcat') do
  it { should be_enabled   }
  it { should be_running   }
end
describe port(8080) do
  it { should be_listening }
end
describe port(8005) do
  it { should be_listening }
end


cookbook/recipes/tomcat.rb
#
# Cookbook Name:: cb-tomcat
# Recipe:: tomcat
#

# Create tomcat group
group node['tomcat']['group']

# Create tomcat user
user node['tomcat']['user'] do
  supports :manage_home => true
  group node['tomcat']['group']
  system true
  home '/home/tomcat'
  shell '/bin/bash'
  comment 'Tomcat user'
end

# Download and unpack tomcat
download_url =
  'http://archive.apache.org/dist/tomcat/tomcat-7/'\
  "v#{node['tomcat']['version']}/bin/apache-tomcat-"\
  "#{node['tomcat']['version']}.tar.gz"
extras_download_url =
  'http://archive.apache.org/dist/tomcat/tomcat-7/'\
  "v#{node['tomcat']['version']}/bin/extras"
ark 'tomcat' do
  url download_url
  checksum 'f0316c128881c4df384771dc0da8f8e80d861385798e57d22fd4068f48ab8724'
  version node['tomcat']['version']
  prefix_root node['tomcat']['install_directory']
  prefix_home node['tomcat']['install_directory']
  owner node['tomcat']['user']
end
node.normal['tomcat']['home_directory'] =
  "#{node['tomcat']['install_directory']}/tomcat"

# Undeploy default webapps
webapps_undeploy = [
  "#{node['tomcat']['install_directory']}/tomcat/webapps/docs",
  "#{node['tomcat']['install_directory']}/tomcat/webapps/examples"
]
if node['tomcat']['webapps']['manager']['deploy'] == false
  webapps_undeploy = webapps_undeploy.concat([
    "#{node['tomcat']['install_directory']}/tomcat/webapps/manager",
    "#{node['tomcat']['install_directory']}/tomcat/webapps/host-manager"
  ])
end
webapps_undeploy.each do |dir|
  directory dir do
    recursive true
    action :delete
  end
end

# Add the version file
template "#{node['tomcat']['install_directory']}/tomcat/VERSION" do
  source 'VERSION.erb'
  mode '644'
  notifies :restart, 'service[tomcat]', :delayed
end

# Download an add JmxRemoteLifecycleListener
# http://tomcat.apache.org/tomcat-7.0-doc/config/listeners.html
jmx_remote_jar =
  "#{node['tomcat']['install_directory']}/tomcat/"\
  'lib/catalina-jmx-remote.jar'
remote_file jmx_remote_jar do
  source "#{extras_download_url}/catalina-jmx-remote.jar"
  mode '644'
  notifies :restart, 'service[tomcat]', :delayed
end

# Update CATALINA_OPTS
node.normal['tomcat']['catalina_opts'] =
  node['tomcat']['catalina_opts'].concat([
    '-Dcom.sun.management.jmxremote=',
    '-Djava.rmi.server.hostname'\
      "=#{node['tomcat']['jmx']['rmi_bind_address']}",
    '-Dcom.sun.management.jmxremote.ssl=false',
    '-Dcom.sun.management.jmxremote.ssl=false',
    '-Dcom.sun.management.jmxremote.authenticate=false'
  ])

# Enable JDWP
# http://docs.oracle.com/javase/1.5.0/docs/guide/jpda/jdwp-spec.html
if node['tomcat']['jdwp']['enabled']
  if node['tomcat']['jdwp']['suspend']
    node.normal['tomcat']['catalina_opts'] =
      node['tomcat']['catalina_opts'].concat([
        '-Xdebug',
        '-Xrunjdwp:transport=dt_socket,address='\
          "#{node['tomcat']['jdwp']['port']},suspend=y,server=y"
      ])
  else
    node.normal['tomcat']['catalina_opts'] =
      node['tomcat']['catalina_opts'].concat([
        '-Xdebug',
        '-Xrunjdwp:transport=dt_socket,address='\
          "#{node['tomcat']['jdwp']['port']},suspend=n,server=y"
      ])
  end
end

# Enable the tomcat service
template '/etc/init.d/tomcat' do
  source 'tomcat.erb'
  mode '755'
  notifies :restart, 'service[tomcat]', :delayed
end

# Tomcat catalina configuration
template "#{node['tomcat']['install_directory']}/tomcat/bin/catalina.sh" do
  source 'catalina.sh.erb'
  mode '755'
  notifies :restart, 'service[tomcat]', :delayed
end

# Tomcat users configuration
tomcat_users_file =
  "#{node['tomcat']['install_directory']}/tomcat/"\
  'conf/tomcat-users.xml'
template tomcat_users_file do
  source 'tomcat-users.xml.erb'
  mode '644'
  notifies :restart, 'service[tomcat]', :delayed
end

# server.xml
template "#{node['tomcat']['install_directory']}/tomcat/conf/server.xml" do
  source 'server.xml.erb'
  mode '644'
  notifies :restart, 'service[tomcat]', :delayed
end

# Enable the tomcat service
service 'tomcat' do
  supports :restart => true, :start => true, :stop => true
  action [:enable]
end

# Install APR
bash 'reload profile prior to apr installation' do
  code 'source /etc/profile'
end
apr_download_url =
  'http://archive.apache.org/dist/apr/'\
  "apr-#{node['tomcat']['apr']['version']}.tar.gz"
apr_util_download_url =
  'http://archive.apache.org/dist/apr/'\
  "apr-util-#{node['tomcat']['apr-util']['version']}.tar.gz"
ark 'apr' do
  url apr_download_url
  checksum '94b1c9d9835cc9e902838b95d62ecc9a39b698f23e3e706812ec65a78ba41af7'
  version node['tomcat']['apr']['version']
  prefix_root node['tomcat']['apr']['install_directory']
  prefix_home node['tomcat']['apr']['install_directory']
  owner 'root'
  action :configure
end
bash 'make apr' do
  user 'root'
  cwd node['tomcat']['apr']['install_directory']
  code <<-EOH
  cd apr-#{node['tomcat']['apr']['version']}
  make && make install
  EOH
end
ark 'apr-util' do
  url apr_util_download_url
  checksum '76db34cb508e346e3bf69347c29ed1500bf0b71bcc48d54271ad9d1c25703743'
  version node['tomcat']['apr-util']['version']
  prefix_root node['tomcat']['apr']['install_directory']
  prefix_home node['tomcat']['apr']['install_directory']
  autoconf_opts ['--with-apr=/usr/local/apr/']
  owner 'root'
  action :configure
end
ark 'tomcat-native' do
  url "file://#{node['tomcat']['install_directory']}/"\
      'tomcat/bin/tomcat-native.tar.gz'
  version node['tomcat']['version']
  prefix_root node['tomcat']['apr']['install_directory']
  prefix_home node['tomcat']['apr']['install_directory']
  strip_components 3
  autoconf_opts ['--with-apr=/usr/local/apr/']
  owner 'root'
  action :configure
end
bash 'make apr utils and tomcat native' do
  user 'root'
  cwd node['tomcat']['apr']['install_directory']
  code <<-EOH
  cd apr-util-#{node['tomcat']['apr-util']['version']}
  make && make install
  cd #{node['tomcat']['apr']['install_directory']}
  cd tomcat-native-#{node['tomcat']['version']}
  make && make install
  EOH
end
template '/etc/profile.d/apr.sh' do
  source 'apr.sh.erb'
  mode '755'
  notifies :restart, 'service[tomcat]', :delayed
end
bash 'reload profile' do
  code 'source /etc/profile'
end


cookbook/attributes/default.rb

# Basic installation
node.default['tomcat']['user'] = 'tomcat'
node.default['tomcat']['group'] = 'tomcat'
node.default['tomcat']['version'] = '7.0.54'
node.default['tomcat']['install_directory'] = '/opt'
node.default['tomcat']['java_opts'] = [
  '-Djava.awt.headless=true',
  '-Dfile.encoding=UTF-8',
  '-server',
  '-XX:-HeapDumpOnOutOfMemoryError',
  '-XX:-OmitStackTraceInFastThrow',
  '-Djava.net.preferIPv4Stack=true'
]
node.default['tomcat']['catalina_opts'] = [
]

# APR
node.default['tomcat']['apr']['version'] = '1.5.1'
node.default['tomcat']['apr-util']['version'] = '1.5.3'
node.default['tomcat']['apr']['install_directory'] = '/usr/local'

# Manager
node.default['tomcat']['webapps']['manager']['deploy'] = true
node.default['tomcat']['webapps']['manager']['admin-script']['enabled'] = true
node.default['tomcat']['webapps']['manager']['admin-script']['pw'] = 'script'
node.default['tomcat']['webapps']['manager']['admin-gui']['enabled'] = true
node.default['tomcat']['webapps']['manager']['admin-gui']['pw'] = 'gui'

# JMX
node.default['tomcat']['jmx']['rmi_registry_port'] = '10001'
node.default['tomcat']['jmx']['rmi_server_port'] = '10002'
node.default['tomcat']['jmx']['rmi_bind_address'] = 'localhost'

# JDWP
node.default['tomcat']['jdwp']['enabled'] = true
node.default['tomcat']['jdwp']['suspend'] = false
node.default['tomcat']['jdwp']['port'] = '10003'

# server.xml
# Connector options from: http://tomcat.apache.org/tomcat-7.0-doc/config/http.html
node.default['tomcat']['shutdown_port'] = '8005'
node.default['tomcat']['shutdown'] = 'shutdown'
node.default['tomcat']['executor']['max_threads'] = '100'
node.default['tomcat']['executor']['min_spare_threads'] = '10'
node.default['tomcat']['connector']['apr']['enabled'] = true
node.default['tomcat']['connector']['ssl_engine'] = 'off'
node.default['tomcat']['connector']['http']['allow_trace'] = 'false'
node.default['tomcat']['connector']['http']['async_timeout'] = '10000'
node.default['tomcat']['connector']['http']['enable_lookups'] = 'false'
node.default['tomcat']['connector']['http']['max_header_count'] = '100'
node.default['tomcat']['connector']['http']['max_parameter_count'] = '10000'
node.default['tomcat']['connector']['http']['max_post_size'] = '2097152'
node.default['tomcat']['connector']['http']['max_save_post_size'] = '4096'
node.default['tomcat']['connector']['http']['parse_body_methods'] = 'POST'
node.default['tomcat']['connector']['http']['port'] = '8080'
node.default['tomcat']['connector']['http']['protocol'] =
  'org.apache.coyote.http11.Http11AprProtocol'
node.default['tomcat']['connector']['http']['proxy_name'] = ''
node.default['tomcat']['connector']['http']['proxy_port'] =
  node['tomcat']['connector']['http']['port']
node.default['tomcat']['connector']['http']['redirect_port'] = '8443'
node.default['tomcat']['connector']['http']['scheme'] = 'http'
node.default['tomcat']['connector']['http']['secure'] = 'false'
node.default['tomcat']['connector']['http']['uri_encoding'] = 'UTF-8'
node.default['tomcat']['connector']['http']['body_encoding_for_uri'] = 'false'
node.default['tomcat']['connector']['http']['use_ipv_hosts'] = 'false'
node.default['tomcat']['connector']['http']['xpoweredby'] = 'false'
node.default['tomcat']['connector']['http']['accept_count'] = '100'
node.default['tomcat']['connector']['http']['acceptor_thread_count'] = '2'
node.default['tomcat']['connector']['http']['acceptor_thread_priority'] = '5'
node.default['tomcat']['connector']['http']['address'] = ''
node.default['tomcat']['connector']['http']['bind_on_init'] = 'true'
node.default['tomcat']['connector']['http']['compressable_mime_type'] = ''
node.default['tomcat']['connector']['http']['compression'] = 'off'
node.default['tomcat']['connector']['http']['compression_min_size'] = '2048'
node.default['tomcat']['connector']['http']['connection_linger'] = '-1'
node.default['tomcat']['connector']['http']['connection_timeout'] = '20000'
node.default['tomcat']['connector']['http']['connection_upload_timeout'] =
  '300000'
node.default['tomcat']['connector']['http']['disable_upload_timeout'] = 'true'
node.default['tomcat']['connector']['http']['executor'] = 'tomcatThreadPool'
node.default['tomcat']['connector']['http']['executor_kill_timeout'] = ''
node.default['tomcat']['connector']['http']['keep_alive_timeout'] = '5000'
node.default['tomcat']['connector']['http']['max_connections'] = ''
node.default['tomcat']['connector']['http']['max_extension_size'] = '8192'
node.default['tomcat']['connector']['http']['max_http_header_size'] = '8192'
node.default['tomcat']['connector']['http']['max_keep_alive_requests'] = '1'
node.default['tomcat']['connector']['http']['max_threads'] = '200'
node.default['tomcat']['connector']['http']['max_trailer_size'] = '8192'
node.default['tomcat']['connector']['http']['min_spare_threads'] = ''
node.default['tomcat']['connector']['http']['no_compression_user_agents'] = ''
node.default['tomcat']['connector']['http']['processor_cache'] = '200'
node.default['tomcat']['connector']['http']['restricted_user_agents'] = ''
node.default['tomcat']['connector']['http']['server'] = 'server'
node.default['tomcat']['connector']['http']['socket_buffer'] = '9000'
node.default['tomcat']['connector']['http']['ssl_enabled'] = 'false'
node.default['tomcat']['connector']['http']['tcp_no_delay'] = 'true'
node.default['tomcat']['connector']['http']['thread_priority'] = '5'
node.default['tomcat']['connector']['http']['defer_accept'] = 'true'
node.default['tomcat']['connector']['http']['poller_size'] = '8192'
node.default['tomcat']['connector']['http']['poll_time'] = '2000'
node.default['tomcat']['connector']['http']['sendfile_size'] = '1024'
node.default['tomcat']['connector']['http']['thread_priority'] = '5'
node.default['tomcat']['connector']['http']['use_comet'] = 'false'
node.default['tomcat']['connector']['http']['use_sendfile'] = 'false'
